plugins/amqp_trx_plugin/amqp_trx_plugin.cpp

```diff
@@ -0,0 +1,188 @@
+         }
+         return;
+      }
+
+      trx_in_progress_size += trx->get_estimated_size();
+      app().post( priority::medium_low, [my=shared_from_this(), trx{std::move(trx)}]() {
+         my->chain_plug->accept_transaction( trx,
+            [my, trx](const fc::static_variant<fc::exception_ptr, chain::transaction_trace_ptr>& result) mutable {
+               // publish to trace plugin as execptions are not reported via controller signal applied_transaction
+               if( result.contains<chain::exception_ptr>() ) {
+                  auto& eptr = result.get<chain::exception_ptr>();
+                  if( my->trace_plug ) {
+                     my->trace_plug->publish_error( trx->id().str(), eptr->code(), eptr->to_string() );
+                  }
+                  dlog( "accept_transaction ${id} exception: ${e}", ("id", trx->id())("e", eptr->to_string()) );
+               } else {
+                  dlog( "accept_transaction ${id}", ("id", trx->id()) );
+               }
+               my->trx_in_progress_size -= trx->get_estimated_size();
+
+            } );
+         } );
+   }
+};
+
+amqp_trx_plugin::amqp_trx_plugin()
+: my(std::make_shared<amqp_trx_plugin_impl>()) {}
+
+amqp_trx_plugin::~amqp_trx_plugin() {}
+
+void amqp_trx_plugin::set_program_options(options_description& cli, options_description& cfg) {
+   auto op = cfg.add_options();
+   op("amqp-trx-address", bpo::value<std::string>(),
+      "AMQP address: Format: amqp://USER:PASSWORD@ADDRESS:PORT\n"
+      "Will consume from 'trx' queue.");
+}
+
+void amqp_trx_plugin::plugin_initialize(const variables_map& options) {
+   try {
+      my->chain_plug = app().find_plugin<chain_plugin>();
+      EOS_ASSERT( my->chain_plug, chain::missing_chain_plugin_exception, "chain_plugin required" );
+
+      my->trace_plug = app().find_plugin<amqp_trace_plugin>(); // optional
+
+      EOS_ASSERT( options.count("amqp-trx-address"), chain::plugin_config_exception, "amqp-trx-address required" );
+      my->amqp_trx_address = options.at("amqp-trx-address").as<std::string>();
+   }
+   FC_LOG_AND_RETHROW()
+}
+
+void amqp_trx_plugin::plugin_startup() {
+   handle_sighup();
+   try {
+
+      if( !my->trace_plug ||
+          ( my->trace_plug->get_state() != abstract_plugin::initialized &&
+            my->trace_plug->get_state() != abstract_plugin::started ) ) {
+         dlog( "running without amqp_trace_plugin" );
+         my->trace_plug = nullptr;
+      } else {
+         // always want trace plugin running if specified so traces can be published
+         my->trace_plug->plugin_startup();
+      }
+
+      ilog( "Starting amqp_trx_plugin" );
+
+      my->thread_pool.emplace( "amqp_t", 1 );
+
+      my->amqp_trx.emplace( my->thread_pool->get_executor(), my->amqp_trx_address, "trx", [](const std::string& err) {
+         elog( "amqp error: ${e}", ("e", err) );
+         app().quit();
+      });
+
+      auto& consumer = my->amqp_trx->consume();
```

Comment:
This needs to be run in the amqp thread.

Reply:
Refactored to run in amqp thread.